<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta name="generator" content="Pod::HtmlEasy/1.1.10 Pod::Parser/1.37 Perl/5.010001 [linux]">
<title>
asarp
</title>
<link rel="stylesheet" href="default.css" type="text/css">
</head>
<body  alink="#FF0000" bgcolor="#FFFFFF" link="#000000" text="#000000" vlink="#000066">
<div class="toc">
<ul>
<li><a href='#NAME'>NAME</a></li>
<li><a href='#SYNOPSIS'>SYNOPSIS</a></li>
<ul>
<li><a href='#OUTPUTS'>OUTPUTS</a></li>
<li><a href='#-I USAGE'>-I USAGE</a></li>
</ul>
<li><a href='#DESCRIPTION'>DESCRIPTION</a></li>
<ul>
<li><a href='#OVERVIEW'>OVERVIEW</a></li>
<li><a href='#TERMINOLOGY'>TERMINOLOGY</a></li>
</ul>
<li><a href='#REQUIREMENT'>REQUIREMENT</a></li>
<li><a href='#SEE ALSO'>SEE ALSO</a></li>
<li><a href='#COPYRIGHT'>COPYRIGHT</a></li>
<li><a href='#AUTHOR'>AUTHOR</a></li>
</ul>
</div>
<div class='pod'>
<a name='NAME'></a><h1>NAME</h1>
<p>asarp.pl -- The new and improved ASARP pipeline to discover ASE/ASARP genes/SNVs, which now supports strand-specific RNA-Seq data.</p>
<p>For details of the older version, refer to the paper:
<i>Li G, Bahn JH, Lee JH, Peng G, Chen Z, Nelson SF, Xiao X. Identification of allele-specific alternative mRNA processing via transcriptome sequencing, Nucleic Acids Research, 2012, 40(13), e104</i> and its Supplementary Materials. Link: <a href='http://nar.oxfordjournals.org/content/40/13/e104' target='_blank'>nar.oxfordjournals.org</a></p>
<p><img src='img/Intro.png' border=0></p>
<a name='SYNOPSIS'></a><h1>SYNOPSIS</h1>
<pre> perl asarp.pl output_file config_file [optional: parameter_file] </pre>
<p><b>NEW</b>: the ASARP pipeline now supports strand-specific RNA-Seq data (which can be processed by the new pre-processing script: <i><a href='procReads.html'>procReads</a></i>. One can set the optional strand-specific flag in the cnofig file. IMPORTANT: the strand-specific option does not work correctly on non-strand-specific data.</p>
<p>ARGUMENTS:</p>
<pre> config_file		the input configuration file which contains all the input file keys and their paths</pre>
<p>OPTIONAL:</p>
<pre> parameter_file		the parameter configuration file which contains all the thresholds and cutoffs
 			if not input, the default.param file in the ASARP main program directory will be used</pre>
<p>Details of the input config and parameter files can be found in the <i><a href='Files.html'>Files</a></i> page. For preparation of the input files used in <code>config_file</code>, see the pre-processing section: <i><a href='rmDup.html'>rmDup</a></i>, <i><a href='mergeSam.html'>mergeSam</a></i>, <i><a href='procReads.html'>procReads</a></i></p>
<a name='OUTPUTS'></a><h2>OUTPUTS</h2>
<p><code>output_file</code> is where the ASARP result summary is output, and meanwhile there will be 4 addtional detailed result files output:</p>
<ul>
<li><a name='output_file.ase.prediction -- the detailed results of (whole-gene-level) ASE patterns (exclusive to other ASARP patterns: AI, AS or AT)'></a><code>output_file.ase.prediction</code>
-- the detailed results of (whole-gene-level) ASE patterns (exclusive to other ASARP patterns: AI, AS or AT)</li>
<li><a name='output_file.gene.prediction -- the detailed results of ASARP results (ASE patterns excluded) arranged by genes'></a><code>output_file.gene.prediction</code>
-- the detailed results of ASARP results (ASE patterns excluded) arranged by genes</li>
<li><a name='output_file.snv.prediction -- the detailed results of ASARP results (ASE patterns excluded) of each individual SNV'></a><code>output_file.snv.prediction</code>
-- the detailed results of ASARP results (ASE patterns excluded) of each individual SNV</li>
<li><a name='output_file.controlSNV.prediction -- the control SNV information of each individual ASARP SNV'></a><code>output_file.controlSNV.prediction</code>
-- the control SNV information of each individual ASARP SNV</li>
</ul>
<a name='-I USAGE'></a><h2>-I USAGE</h2>
<p>Because asarp.pl requires other perl files in the same folder to run, <code>-I path</code> can be used if one would like to run ASARP elsewhere by adding its <code>path</code>.</p>
<pre> perl -I path path/asarp.pl output_file config_file parameter_file</pre>
<p>Note that in such a case, one should be careful of the locations of the config and parameter files. Abosulute paths are suggested for the files in <code>config_file</code>.</p>
<a name='DESCRIPTION'></a><h1>DESCRIPTION</h1>
<p>The ASARP method is presented below:</p>
<a name='OVERVIEW'></a><h2>OVERVIEW</h2>
<p>The procedures (rules) for ASARP are illustrated in the following figure and terminology explained below:</p>
<p><img src='img/ASARP_core.png' border=0></p>
<p>There are basically 3 steps.</p>
<p>1. parse the input files and compile alternative mRNA processing events. see outputs of <i><a href='procReads.html'>procReads</a></i></p>
<p>2. get the SNVs and match them with the events.</p>
<p>3. process ASARP (including ASE) patterns and output the formatted results.</p>
<a name='TERMINOLOGY'></a><h2>TERMINOLOGY</h2>
<p>The predictions that ASARP makes are desribed below:</p>
<p>Allele-Specific Expression (ASE)</p>
<ul>
<li><a name='ASE: a single SNV is called to have an ASE pattern (or simply ASE SNV) if its allelic ratio significally diverges from 0.5 (in other words 1:1 for Ref:Alt).'></a><b>ASE</b>: a single SNV is called to have an ASE pattern (or simply ASE SNV) if its allelic ratio significally diverges from 0.5 (in other words 1:1 for Ref:Alt).</li>
</ul>
<p>Allele-Specific Alternative RNA Processing (<b>ASARP</b>) types:</p>
<ul>
<li><a name='ASAS: Allele-Specific Alternative Splicing;'></a><b>ASAS</b>: Allele-Specific Alternative Splicing;</li>
<li><a name='ASAI: Allele-Specific Alternative (5'-end) Initiation;'></a><b>ASAI</b>: Allele-Specific Alternative (5'-end) Initiation;</li>
<li><a name='ASAT: Allele-Specific Alternative (3'-end) Termination, or Alternative Polyadenylation'></a><b>ASAT</b>: Allele-Specific Alternative (3'-end) Termination, or Alternative Polyadenylation</li>
</ul>
<p>How to categorize ASARP patterns into specific Allele-Specific AI/AS/AT and/or combinations of them depends on whether the candidate SNV locations are in internal exons/introns (AS) and/or alternative 3' or 5' UTRs (AI/AT). A complex ASARP gene is with ASARP SNVs in more than one categories.</p>
<p><img src='img/Types.png' border=0></p>
<p><b>NEV</b>: Normalized Expression Value, a PSI (Percent Spliced-In) like value to measure whether an event (also alternatively processed region) is also alternatively processed according to RNA-Seq (gene expression). It is calculated as (note that in some events only <code>NEV_gene</code> is available)</p>
<ul>
<li><a name='NEV_sp = min (NEV_flanking, NEV_gene), where'></a><code>NEV_sp = min (NEV_flanking, NEV_gene)</code>, where</li>
<li><a name='NEV_flanking = (# event_reads/event_length)/(# flanking_region_total_reads/flanking_region_total_length), and'></a><code>NEV_flanking = (# event_reads/event_length)/(# flanking_region_total_reads/flanking_region_total_length)</code>, and</li>
<li><a name='NEV_gene = (# event_reads/event_length)/(# gene_constitutive_exon_reads/gene_constitutive_exon_length)'></a><code>NEV_gene = (# event_reads/event_length)/(# gene_constitutive_exon_reads/gene_constitutive_exon_length)</code></li>
</ul>
<p><code>*_length</code> means the total number of positions within the * region with non-zero reads.</p>
<p><img src='img/Event.png' border=0></p>
<a name='REQUIREMENT'></a><h1>REQUIREMENT</h1>
<p><code>Statistics::R</code>: has to be installed. See <a href='http://search.cpan.org/~fangly/Statistics-R/lib/Statistics/R.pm' target='_blank'>search.cpan.org</a></p>
<a name='SEE ALSO'></a><h1>SEE ALSO</h1>
<p><i><a href='Overview.html'>Overview</a></i>, <i><a href='fileParser.html'>fileParser</a></i>, <i><a href='snpParser.html'>snpParser</a></i>, <i><a href='MyConstants.html'>MyConstants</a></i></p>
<a name='COPYRIGHT'></a><h1>COPYRIGHT</h1>
<p>This pipeline is free software; you can redistribute it and/or modify it given that the related works and authors are cited and acknowledged.</p>
<p>This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose.</p>
<a name='AUTHOR'></a><h1>AUTHOR</h1>
<p>Cyrus Tak-Ming CHAN</p>
<p>Xiao Lab, Department of Integrative Biology &amp; Physiology, UCLA</p>
</div></body></html>
