<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta name="generator" content="Pod::HtmlEasy/1.1.10 Pod::Parser/1.37 Perl/5.010001 [linux]">
<title>
mergeSnvs
</title>
<link rel="stylesheet" href="default.css" type="text/css">
</head>
<body  alink="#FF0000" bgcolor="#FFFFFF" link="#000000" text="#000000" vlink="#000066">
<div class="toc">
<ul>
<li><a href='#NAME'>NAME</a></li>
<li><a href='#SYNOPSIS'>SYNOPSIS</a></li>
<li><a href='#EXAMPLE OUTPUTS'>EXAMPLE OUTPUTS</a></li>
<li><a href='#SEE ALSO'>SEE ALSO</a></li>
<li><a href='#COPYRIGHT'>COPYRIGHT</a></li>
<li><a href='#AUTHOR'>AUTHOR</a></li>
</ul>
</div>
<div class='pod'>
<a name='NAME'></a><h1>NAME</h1>
<p>mergeSnvs.pl -- merge SNVs of different chromosomes into one SNV list
the equivalent can be achieved using cat and grep commands
for strand-specific cases '+$' and '\-$' can get all SNVs from the
+ and - strands respectively</p>
<p>After the full pre-processing, one can run ASARP for each chromosome's
SNVs, or merge all chromosomes' SNVs into a single SNV list using this
script (equivalent to using Unix cat and/or grep).</p>
<a name='SYNOPSIS'></a><h1>SYNOPSIS</h1>
<p>USAGE:</p>
<pre> perl mergeSnvs.pl  prefix suffix mono=x output [isStrandSp]</pre>
<p>ARGUMENTS:</p>
<pre> prefix		the folder path and prefix string for all chr* SNV files
 		e.g. &quot;/home/N.A+/&quot; for all chr* SNVs in that folder
 suffix		the suffix after chr[1..22/X/Y/M]
 mono=		options (no space): 
		&quot;mono=0&quot;: allow mono-allelic SNVs; 
		&quot;mono=x&quot;: discard SNVs with any allele &lt; x reads; 
		useful when genotype is unknown and dbSNPs are used
		where mono-allelic SNVs would be called
 output		the output file name for the merged SNVs
 		note that output.plus and output.minus are by-product
 		outputs for + and - strands for strand-specific cases
 		make sure the output and chr* SNV file names do not conflict</pre>
<p>OPTIONAL:</p>
<pre> isStrandSp	set 1 if the SNV files are strand-specific
		default: non-strand-specific</pre>
<p>Scatter plots: output.scatt.pdf and histogram: output.hist.pdf will be generated alongside the output, visualizing the SNV statistics.</p>
<a name='EXAMPLE OUTPUTS'></a><h1>EXAMPLE OUTPUTS</h1>
<p>Examples of scatter plots and histograms (mono=10):</p>
<p><img src='img/snvs.png' border=0></p>
<a name='SEE ALSO'></a><h1>SEE ALSO</h1>
<p><i><a href='rmDup.html'>rmDup</a></i>, <i><a href='mergeSam.html'>mergeSam</a></i>, <i><a href='procReads.html'>procReads</a></i>, <i><a href='procReadsJ.html'>procReadsJ</a></i>, <i><a href='asarp.html'>asarp</a></i></p>
<a name='COPYRIGHT'></a><h1>COPYRIGHT</h1>
<p>This pipeline is free software; you can redistribute it and/or modify it given that the related works and authors are cited and acknowledged.</p>
<p>This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose.</p>
<a name='AUTHOR'></a><h1>AUTHOR</h1>
<p>Cyrus Tak-Ming CHAN</p>
<p>Xiao Lab, Department of Integrative Biology &amp; Physiology, UCLA</p>
</div></body></html>
