<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta name="generator" content="Pod::HtmlEasy/1.1.10 Pod::Parser/1.37 Perl/5.010001 [linux]">
<title>
procReads
</title>
<link rel="stylesheet" href="default.css" type="text/css">
</head>
<body  alink="#FF0000" bgcolor="#FFFFFF" link="#000000" text="#000000" vlink="#000066">
<div class="toc">
<ul>
<li><a href='#NAME'>NAME</a></li>
<li><a href='#SYNOPSIS'>SYNOPSIS</a></li>
<li><a href='#DESCRIPTION'>DESCRIPTION</a></li>
<ul>
<li><a href='#INPUT'>INPUT</a></li>
<li><a href='#OUTPUT'>OUTPUT</a></li>
</ul>
<li><a href='#SEE ALSO'>SEE ALSO</a></li>
<li><a href='#COPYRIGHT'>COPYRIGHT</a></li>
<li><a href='#AUTHOR'>AUTHOR</a></li>
</ul>
</div>
<div class='pod'>
<a name='NAME'></a><h1>NAME</h1>
<p>procReads.pl -- Processing a duplicate-removed SAM file (<i><a href='rmDup.html'>rmDup</a></i>) of a chromosome to generate the chromosome specific SNV list and the bedgraph file. The output files are used as input files for the ASARP pipeline.</p>
<p>The new procReads.pl supports strand-specific paired-end RNA-Seq data (i.e. the SAM file strand information is reliable) for more accurate results.</p>
<a name='SYNOPSIS'></a><h1>SYNOPSIS</h1>
<p>This is part of the full pre-processing:</p>
<ul>
<p>1. rmDup (removing PCR duplicates for SAM files (including Dr. JH Lee's SAM format); samtools/bedtools can be used for standard SAM files)</p>
<p>2. mergeSam (merging SAM files if there are independent duplicates)</p>
<p>3. <b>procReads</b> (processing SAM files to get SNV read counts and generate bedgraph files)</p>
</ul>
<p>USAGE:</p>
<pre> perl procReads.pl input_sam_file input_snvs output_snvs output_bedgraph is_paired_end [discarded_read_pos]</pre>
<p>NOTE:</p>
<p>the read processing script is for a standard SAM file with
11 attributes. You can also samtools and bedtools; or
to use procReadsJ.pl on the special jsam files introduced
by Dr. Jae-Hyung Lee for RNA-editin and allele specific
expression (ASE) studies</p>
<p>ARGUMENTS:</p>
<pre> chr			chromosome to be investigated (correspond to the input_sam_file)
 input_sam_file		SAM file input after duplicate removal (use rmDup.pl)
 intput_snvs		input SNV list (without read counts)
 output_snvs		output SNV candidates with read counts
 output_bedgraph	output bedgraph file, see below for the details:
			<a href='http://genome.ucsc.edu/goldenPath/help/bedgraph.html' target='_blank'>genome.ucsc.edu</a>
 is_paired_end		0: single-end; 1: paired-end
			For paired-end reads, all reads should be paired up, 
			where pair-1 should be always followed by pair-2 in the next line.</pre>
<p>OPTIONAL [strongly recommended to be input]:</p>
<pre> is_strand_sp		0: non-strand specific (or no input); 
			1: strand-specific with pair 1 **sense**;
			2: strand-specific with pair 1 **anti-sense**
			Be careful with the strand-specific setting as it will give totally opposite 
			strand information if wrongly set.

			The strand-specific option is used for strand-specific RNA-Seq data.
			When set, specialized bedgraph files will be output (output_bedgraph)
			where there is a 5th extra attribute specifying the strand: + or -
			besides the standard ones: <a href='http://genome.ucsc.edu/goldenPath/help/bedgraph.html' target='_blank'>genome.ucsc.edu</a>
			One can use grep and cut to get +/- strand only bedgraphs.</pre>
<p>OPTIONAL [if input, must be input in order following is_strand_sp]:</p>
<pre> bedgraph_title		a short title for the output bedgraph files (will be put in description of the header line)
                        if there are spaces in between it should be quoted
			e.g. &quot;nbt.editing reads: distinct after dup removal&quot;
			if not input, &quot;default&quot; will be used as the short title

 discarded_read_pos	masked-out (low-quality) read positions in calculating 
			the max read quality scores, 
			in 1-based, inclusive, interval (a:b,c:d,... no space) format:
			e.g. 1:1,61:70 will discard the 1st, 61st-70th read positions.
			NOTE: the remaining reads will still contain the positions.</pre>
<a name='DESCRIPTION'></a><h1>DESCRIPTION</h1>
<a name='INPUT'></a><h2>INPUT</h2>
<p><code>input_sam_file</code> should contain only 1 chromosome, and it should be in Dr. Jae-Hyung Lee's SAM format (check out <a href='http://www.ncbi.nlm.nih.gov/pubmed/21960545' target='_blank'>www.ncbi.nlm.nih.gov</a> for more details)</p>
<p>The SNP list should be in a format like this:</p>
<pre> chr1 20129 C&gt;T rs12354148 2:0:0
 chr1 118617 T&gt;C na 1:0:0
 chr1 237763 G&gt;A rs79665216 1:0:0
 chr1 565508 G&gt;A rs9283150 0:1:0</pre>
<p>Each line is space separated, with</p>
<pre> chromosome
 location 
 ref_allele&gt;alt_allele 
 dbSnp_id 
 [read_counts ref:alt:others]</pre>
<p>Only the first 4 fields will be parsed so
<code>read_counts</code> are not needed and ignored which are from DNA genomic sequencing.</p>
<p>To avoid unnecessary computational time to read SNVs of other chrosomes than the one in <code>inptu_sam_file</code>, it is suggested to keep SNVs of the same chromosome in one seperate file.</p>
<a name='OUTPUT'></a><h2>OUTPUT</h2>
<p>By default, the strand specific flag <code>is_strand_sp</code> is unset, and the <code>inptu_sam_file</code> strand information is considered unreliable and not used. The output files are standard bedgraph files <a href='http://genome.ucsc.edu/goldenPath/help/bedgraph.html' target='_blank'>genome.ucsc.edu</a> with space as the dilimiter, and SNV files should follow the file format described in <i><a href='Files.html'>Files</a></i>:</p>
<p>Example SNV output lists look like this:</p>
<pre> chr10 1046712 G&gt;A rs2306409 50:39:0
 chr10 1054444 A&gt;G rs11253567 2:0:0
 chr10 1055866 A&gt;T rs4880751 1:2:1
 chr10 1055949 G&gt;A rs12355506 7:2:0
 chr10 1055968 G&gt;A rs72478237 6:5:0
 chr10 1060218 G&gt;A rs3207775 42:37:0</pre>
<p>When <code>is_strand_sp</code> is set, the program outputs bedgraph and SNV files with the extra last strand attributes. Output SNV examples would look like:</p>
<pre> chr10 1046712 G&gt;A rs2306409 30:23:0 +
 chr10 1055866 A&gt;T rs4880751 1:2:0 +
 chr10 1055949 G&gt;A rs12355506 2:0:0 +
 chr10 1055968 G&gt;A rs72478237 2:2:0 +
 chr10 1060218 G&gt;A rs3207775 27:22:0 +
 ...
 chr10 1046712 G&gt;A rs2306409 20:16:0 -
 chr10 1054444 A&gt;G rs11253567 2:0:0 -
 chr10 1055949 G&gt;A rs12355506 5:2:0 -
 chr10 1055968 G&gt;A rs72478237 4:3:0 -
 chr10 1060218 G&gt;A rs3207775 15:15:0 -</pre>
<p>As a result, one SNV may appear twice if it has valid ref:alt:wrnt read counts on both + and - strands. In the example above, one can have more accurate information for the RNA read counts, especially when there are genes on the opposite strands.</p>
<p>The output bedgraph lines would look like:</p>
<pre> chr10 181481 181482 7 +
 chr10 181482 181483 9 +
 chr10 181483 181499 10 +
 ...
 chr10 181479 181482 5 -
 chr10 181482 181483 8 -
 chr10 181483 181499 9 -
 ...</pre>
<p>Note that all + lines are output before any - lines output. While the 5th strand attribute is not specified in the bedgraph standard,
two additional bedgraph files are output, with suffixes .plus and .minus, to provide the + and - only standard bedgraph tracks respectively.</p>
<a name='SEE ALSO'></a><h1>SEE ALSO</h1>
<p><i><a href='rmDup.html'>rmDup</a></i>, <i><a href='mergeSam.html'>mergeSam</a></i>, <i><a href='asarp.html'>asarp</a></i></p>
<a name='COPYRIGHT'></a><h1>COPYRIGHT</h1>
<p>This pipeline is free software; you can redistribute it and/or modify it given that the related works and authors are cited and acknowledged.</p>
<p>This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose.</p>
<a name='AUTHOR'></a><h1>AUTHOR</h1>
<p>Cyrus Tak-Ming CHAN</p>
<p>Xiao Lab, Department of Integrative Biology &amp; Physiology, UCLA</p>
</div></body></html>
