<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta name="generator" content="Pod::HtmlEasy/1.1.10 Pod::Parser/1.37 Perl/5.010001 [linux]">
<title>
snp_distri
</title>
<link rel="stylesheet" href="default.css" type="text/css">
</head>
<body  alink="#FF0000" bgcolor="#FFFFFF" link="#000000" text="#000000" vlink="#000066">
<div class="toc">
<ul>
<li><a href='#NAME'>NAME</a></li>
<li><a href='#SYNOPSIS'>SYNOPSIS</a></li>
<li><a href='#DESCRIPTION'>DESCRIPTION</a></li>
<li><a href='#SEE ALSO'>SEE ALSO</a></li>
<li><a href='#COPYRIGHT'>COPYRIGHT</a></li>
<li><a href='#AUTHOR'>AUTHOR</a></li>
</ul>
</div>
<div class='pod'>
<a name='NAME'></a><h1>NAME</h1>
<p>snp_distri.pl -- To calculate the SNV (position) distribution in exons, UTRs, introns, etc., according to transcript annotations. A SNV is disarded (not contributing to the total SNV number) if it is covered by &lt; 2 RNA-Seq reads.</p>
<p>It also serves as an introductory application script to get familiar with the ASARP pipeline (<i><a href='asarp.html'>asarp</a></i>)</p>
<a name='SYNOPSIS'></a><h1>SYNOPSIS</h1>
<pre>  perl snp_distri.pl output_file snv_file transcript_file strand_flag [powerful_snv_cutoff [pwr_snv_details ordinary_snv_details [is_detailed]]]</pre>
<p>ARGUMENTS:</p>
<p><code>output_file</code>: output file (see below)
<code>snv_file</code>:	SNV input file (can be strand specific, see strand_flag), see <i><a href='Files.html'>Files</a></i>
<code>transcript_file</code>:	gene annotation file (i.e. xiaofile used in ASARP config), see <i><a href='Files.html'>Files</a></i></p>
<p><code>strand_flag</code>:	0--non-strand-specific; 1/2--strand-specific (not distinguishable at this step).
If set, there will be 2 snv_details files for each category (pwr or ordinary),
with .plus and .minus indicating the distribtuion in the + and - strands</p>
<p>OPTIONAL:</p>
<p>The optional arguments must be input in order.</p>
<p><code>powerful_snv_cutoff</code>: an optional cutoff to categorize SNVs into powerful (&gt;= <code>powerful_snv_cutoff</code>) and non-powerful types. Default: 20</p>
<p><code>pwr_snv_details</code></p>
<p><code>ordinary_snv_details</code>:
These two are the output files for the detailed SNV categories of
powerful and non-powerful (ordinary) SNVs respectively.</p>
<p><code>is_detailed</code>:	If is_detailed is set 1 after both *_snv_details, detailed information of the
SNVs will be output: chr;geneName;snvPos;type
=over 6</p>
<li><a name='An example transcript_file is ../data/hg19.merged.to.ensg.all.tx.03.18.2011.txt, which was created by merging ensembl Refseq, UCSC knowngene, Gencode gene, and Vegagene.'></a>An example transcript_file is ../data/hg19.merged.to.ensg.all.tx.03.18.2011.txt,
which was created by merging ensembl Refseq, UCSC knowngene, Gencode
gene, and Vegagene.</li>
</ul>
<p><code>output_file</code>: Tab dilimited counts of SNV positions in different gene regions. Headers are included as illustrated in the following example:</p>
<pre>	Type    Exon    Intron  5'UTR   3'UTR   Complex In-gene Intergenic      Total
	Powerful(&gt;=20)  3788    972     586     4097    1464    7937    342     8279
	Non-powerful(&lt;20)       3846    32143   1246    3679    1323    39566   6580    46146
	Overall 7634    33115   1832    7776    2787    47503   6922    54425</pre>
<p><b>NOTE: SNVs with only one read count (e.g. 1:0:x or 0:1:x) will be discarded by all SNV analysis components in the pipeline, so they are not included, even in Non-powerful categories</b></p>
<a name='DESCRIPTION'></a><h1>DESCRIPTION</h1>
<p>The application is SNV position (ref genome location) oriented. In other words, if (rarely, and not seen in our data) one position contains multiple SNVs, it will be still considered as one SNV (position). A SNV may overalp multiple transcripts of multiple genes. The rule to determine the categories of a SNV is as follows:</p>
<p>If the SNV overlaps certain transcript exon blocks, it is considered as in (coding) exons, regardless the times it overlaps.</p>
<p>If the SNV overlaps certain 5'/3' UTRs, it is considered as in 5'/3' UTRs (non-coding), regardless the times it overlaps.</p>
<p>Only when a SNV never overlaps any exons nor 5'/3' UTRs, and its genome location is within certain transcript span, it is considered as in introns, regardless of the times it overlaps. As a result, intron SNVs are exclusive to exons and 5'/3' UTRs.</p>
<p>A SNV can be categorized into multiple categories, denoted as complex. Therefore, complex contains SNVs with combinations of Exon and 5'/3' UTR types. In-gene SNVs are the union of all the categories: exon, 5' UTR, 3' UTR, intron. Therefore, the sum of intron, exon, 5' and 3' UTR SNV counts will be larger than the total in-gene SNV count if complex SNVs exist. In-gene and Intergenic SNVs are exclusive to each other.</p>
<p>For any of the above cases, a SNV is considered in-gene. If a SNV is not in-gene at all, it is considered in the intergenic regions.</p>
<p>The application also outputs percentage (over the total SNV positions) sumamries for powerful, non-powerful and overall SNV distributions. Sample output:</p>
<pre>	Calculating powerful SNV distribution...
	SNV Position Distribution:
	Exon: 3788 (45.75%)
	Intron: 972 (11.74%)
	5' UTR: 586 (7.08%)
	3' UTR: 4097 (49.49%)
	Complex (in-gene): 1464 (17.68%)
	In-gene total: 7937 (95.87%)
	Intergenic total: 342 (4.13%)
	Total SNV Positions: 8279

	Calculating non-powerful SNV distribution...
	SNV Position Distribution:
	Exon: 3846 (8.33%)
	Intron: 32143 (69.66%)
	5' UTR: 1246 (2.70%)
	3' UTR: 3679 (7.97%)
	Complex (in-gene): 1323 (2.87%)
	In-gene total: 39566 (85.74%)
	Intergenic total: 6580 (14.26%)
	Total SNV Positions: 46146
	...</pre>
<p>If <code>pwr_snv_details</code> and <code>ordinary_snv_details</code> are input, you will have two addtional output files providing detailed categories of all individual <b>in-gene</b> SNVs. E.g.</p>
<pre> chr	pos	category	gene	[strand]
 chr1    6583701 3'UTR;  NOL9    -
 chr1    223967953       3'UTR;  TP53BP2 -
 chr1    225607144       EXON;   LBR     -

 chr1    10366692        3'UTR;  KIF1B   +
 chr1    244845185       INTRON; FAM152A +
 chr1    1563789 EXON;   MIB2    +</pre>
<a name='SEE ALSO'></a><h1>SEE ALSO</h1>
<p><i><a href='asarp.html'>asarp</a></i>, <i><a href='fileParser.html'>fileParser</a></i>, <i><a href='snpParser.html'>snpParser</a></i>, <i><a href='MyConstants.html'>MyConstants</a></i></p>
<a name='COPYRIGHT'></a><h1>COPYRIGHT</h1>
<p>This pipeline is free software; you can redistribute it and/or modify it given that the related works and authors are cited and acknowledged.</p>
<p>This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose.</p>
<a name='AUTHOR'></a><h1>AUTHOR</h1>
<p>Cyrus Tak-Ming CHAN</p>
<p>Xiao Lab, Department of Integrative Biology &amp; Physiology, UCLA</p>
</div></body></html>
