<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta name="generator" content="Pod::HtmlEasy/1.1.10 Pod::Parser/1.37 Perl/5.014002 [linux]">
<title>
snpParser
</title>
<link rel="stylesheet" href="default.css" type="text/css">
</head>
<body  alink="#FF0000" bgcolor="#FFFFFF" link="#000000" text="#000000" vlink="#000066">
<div class="toc">
<ul>
<li><a href='#NAME'>NAME</a></li>
<li><a href='#SYNOPSIS'>SYNOPSIS</a></li>
<li><a href='#REQUIREMENT'>REQUIREMENT</a></li>
<li><a href='#DESCRIPTION'>DESCRIPTION</a></li>
<ul>
<li><a href='#SNV List Format'>SNV List Format</a></li>
<li><a href='#Sub-routines (major)'>Sub-routines (major)</a></li>
</ul>
<li><a href='#SEE ALSO'>SEE ALSO</a></li>
<li><a href='#COPYRIGHT'>COPYRIGHT</a></li>
<li><a href='#AUTHOR'>AUTHOR</a></li>
</ul>
</div>
<div class='pod'>
<a name='NAME'></a><h1>NAME</h1>
<p>snpParser.pl -- All the sub-routines for SNV (sometimes denoted interchangeably as SNP) handling in the ASARP pipeline.</p>
<a name='SYNOPSIS'></a><h1>SYNOPSIS</h1>
<pre>	use Statistics::R; #interact with R
	require &quot;fileParser.pl&quot;; #sub's for input annotation files
	require &quot;snpParser.pl&quot;; #sub's for snps
	...</pre>
<p>... get all configs, input files (see <i><a href='fileParser.html'>fileParser</a></i>)</p>
<pre>
	# read and parse SNVs
	my $snpRef = initSnp($snpF, $POWCUTOFF);
	# match SNVs with gene transcript annotations
	my $geneSnpRef = setGeneSnps($snpRef, $transRef);
	# match gene SNVs with AI/AT and alternative splicing (AS) events
	my ($snpEventsRef) = 
	setSnpEvents($geneSnpRef, $altRef, $splicingRef);

	# calculate NEV and filter the matched gene SNVs with AI/AT/AS events
	my ($snpsNevRef) = 
	filterSnpEventsWithNev($snpRef, $geneSnpRef, $snpEventsRef, $bedF, 
	$allEventsListRef, $NEVCUTOFFLOWER, $NEVCUTOFFUPPER); 

	# process ASE and ASARP
	my ($allAsarpsRef) = 
	processASEWithNev($snpRef, $geneSnpRef, $snpsNevRef, $SNVPCUTOFF, 
	$ASARPPCUTOFF, $ALRATIOCUTOFF);

	# format results to output
	my $outputGene = $outputFile.'.gene.prediction';
	outputRawASARP($allAsarpsRef, 'ASARPgene', $outputGene);
	my $allNarOutput = formatOutputVerNAR($allAsarpsRef);</pre>
<a name='REQUIREMENT'></a><h1>REQUIREMENT</h1>
<p><code>Statistics::R</code>: has to be installed. See <a href='http://search.cpan.org/~fangly/Statistics-R/lib/Statistics/R.pm' target='_blank'>search.cpan.org</a></p>
<a name='DESCRIPTION'></a><h1>DESCRIPTION</h1>
<p>This perl file contains all the sub-routines for SNV handling and ASARP processing, as well as result formatting. They are quite procedural and one should first get the input files such as annotations and events using the sub-routines in <i><a href='fileParser.html'>fileParser</a></i>.</p>
<p>Basically there are 3 steps:</p>
<p>1. read and parse the individual SNVs</p>
<p>2. match the SNVs to transcripts, and then events, and then filter them based on the PSI like Normalized Expression Value (NEV) calculation</p>
<p>3. process the SNVs with ASE patterns and SNV pairs with other ASARP patterns: AI/AT/AS, and output the formatted results</p>
<p>AI/AT/AS categories are briefly illustrated below (where the red dots represent SNVs with ASARP patterns):</p>
<p><img src='img/Types.png' border=0></p>
<a name='SNV List Format'></a><h2>SNV List Format</h2>
<p>The SNV list input file contains the list of all SNVs covered by RNA-Seq in
the genes of interest, with the read counts of the reference (Ref) and alternative (Alt) alleles.</p>
<p>This file is space delimited with the following fields for
each SNV:</p>
<pre>	chromosome
	coordinate
	alleles (reference allele&gt;alternative allele)
	dbsnpID
	RNA-Seq counts 
	(# reads for 
	reference allele:alternative allele:wrong nucleotide)</pre>
<p>Example file: <i><a href='../data/snp.list.fig3' target='_blank'>../data/snp.list.fig3</a></i></p>
<a name='Sub-routines (major)'></a><h2>Sub-routines (major)</h2>
<ul>
<li><a name='initSnp'></a><code>initSnp</code></li>
<p>read and parse SNVs</p>
<pre>  input: ($snpF, $POWCUTOFF) --SNV file path, powerful SNV cutoff
  
  output $snpRef 
  --reference to SNVs, categorized into powerful/non-powerful internally
  
  p-value cutoff (Chi-Squared)</pre>
<li><a name='setGeneSnps'></a><code>setGeneSnps</code></li>
<p>match SNVs with gene transcript annotations</p>
<pre>  input: ($snpRef, $transRef);
  --reference to SNVs, reference to gene transcripts
  
  output: $geneSnpRef 
  --reference to SNVs matching gene transcripts</pre>
<li><a name='setSnpEvents'></a><code>setSnpEvents</code></li>
<p>match gene SNVs with AI/AT and alternative splicing (AS) events</p>
<pre>  input: ($geneSnpRef, $altRef, $splicingRef)
  --reference to gene SNVs ($geneSnpRef), 
  --reference to AI/AT events ($altRef)
  --reference to AS events ($splicingRef)

  output: ($snpEventsRef) --gene SNVs matching AI/AT/AS events</pre>
<li><a name='filterSnpEventsWithNev'></a><code>filterSnpEventsWithNev</code></li>
<p>calculate NEV and filter the matched gene SNVs with AI/AT/AS events</p>
<pre>  input: ($snpRef, $geneSnpRef, $snpEventsRef, $bedF, 
  $allEventsListRef, $NEVCUTOFFLOWER, $NEVCUTOFFUPPER)
  --reference to SNVs ($snpRef),
  --reference to gene SNVs ($geneSnpRef),
  --reference to SNVs matching AI/AT/AS events ($snpEventsRef),
  --reference to bed folder for mapped reads ($bedF),
  --reference to all parsed events ($allEvetnsListRef),
  --lower and upper cutoffs (excl.) for NEV ($NEVCUTOFFLOWER/UPPER)

  output: ($snpsNevRef) --gene SNVs matching NEV and AI/AT/AS events </pre>
<li><a name='processASEWithNev'></a><code>processASEWithNev</code></li>
<p>process ASE and ASARP</p>
<pre>  intput: ($snpRef, $geneSnpRef, $snpsNevRef, 
  $SNVPCUTOFF, $ASARPPCUTOFF, $ALRATIOCUTOFF)
  --see above for $snpRef, $geneSnpRef, $snpsNevRef
  --Chi-Squared Test p-value cutoff on individual SNVs for ASE ($SNVPCUTOFF)
  --Fisher's Exact Test p-value cutoff on target-control SNV pairs for
  ASARP ($ASARPPCUTOFF)
  --allelic ratio difference cutoff for target-control SNV pairs for
  ASARP ($ALRATIOCUTOFF)

  output: ($allAsarpsRef) --reference to all ASE and ASARP results </pre>
<li><a name='outputRawASARP'></a><code>outputRawASARP</code></li>
<p>format results to output</p>
<pre>  input: ($allAsarpsRef, $key, $outputFile)
  --reference to ASARP results ($allAsarpsRef)
  --result type to output ($key) with choices: 
  'ASEgene'--ASE results arranged by genes,
  'ASARPgene'--ASARP results arranged by genes
  'ASARPsnp' --ASARP results arranged by SNVs
  --the output file for the results ($outputFile)

  output: corresponding ASE/ASARP results written to $outputGene</pre>
<li><a name='formatOutputVerNAR'></a><code>formatOutputVerNAR</code></li>
<p>format results to be like the old version for NAR</p>
<pre>  input: $allAsarpsRef --see above
  
  output: ($allNarOutput) 
  --text formatted according to the old version</pre>
<a name='SEE ALSO'></a><h1>SEE ALSO</h1>
<p><i><a href='asarp.html'>asarp</a></i>, <i><a href='fileParser.html'>fileParser</a></i>, <i><a href='MyConstants.html'>MyConstants</a></i></p>
<a name='COPYRIGHT'></a><h1>COPYRIGHT</h1>
<p>This pipeline is free software; you can redistribute it and/or modify it given that the related works and authors are cited and acknowledged.</p>
<p>This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose.</p>
<a name='AUTHOR'></a><h1>AUTHOR</h1>
<p>Cyrus Tak-Ming CHAN</p>
<p>Xiao Lab, Department of Integrative Biology &amp; Physiology, UCLA</p>
</div></body></html>
