<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta name="generator" content="Pod::HtmlEasy/1.1.10 Pod::Parser/1.37 Perl/5.014002 [linux]">
<title>
snpParser
</title>
<link rel="stylesheet" href="default.css" type="text/css">
</head>
<body  alink="#FF0000" bgcolor="#FFFFFF" link="#000000" text="#000000" vlink="#000066">
<div class="toc">
<ul>
<li><a href='#NAME'>NAME</a></li>
<li><a href='#SYNOPSIS'>SYNOPSIS</a></li>
<li><a href='#DESCRIPTION'>DESCRIPTION</a></li>
<li><a href='#SEE ALSO'>SEE ALSO</a></li>
<li><a href='#COPYRIGHT'>COPYRIGHT</a></li>
<li><a href='#AUTHOR'>AUTHOR</a></li>
</ul>
</div>
<div class='pod'>
<a name='NAME'></a><h1>NAME</h1>
<p>snpParser.pl -- All the sub-routines for SNV (sometimes termed SNP internally) handling in the ASARP pipeline.</p>
<a name='SYNOPSIS'></a><h1>SYNOPSIS</h1>
<pre>
	require &quot;fileParser.pl&quot;; #sub's for input annotation files
	require &quot;snpParser.pl&quot;; #sub's for snps
	
	... #get all configs, input files (see L&lt;fileParser&gt;)

	# read and parse SNVs
	my $snpRef = initSnp($snpF, $POWCUTOFF);
	# match SNVs with gene transcript annotations
	my $geneSnpRef = setGeneSnps($snpRef, $transRef);
	# match gene SNVs with AI/AT and alternative splicing (AS) events
	my ($snpEventsRef) = 
	setSnpEvents($geneSnpRef, $altRef, $splicingRef);

	# calculate NEV and filter the matched gene SNVs with AI/AT/AS events
	my ($snpsNevRef) = 
	filterSnpEventsWithNev($snpRef, $geneSnpRef, $snpEventsRef, $bedF, $allEventsListRef, $NEVCUTOFFLOWER, $NEVCUTOFFUPPER); 

	# process ASE and ASARP
	my ($allAsarpsRef) = 
	processASEWithNev($snpRef, $geneSnpRef, $snpsNevRef, $SNVPCUTOFF, $ASARPPCUTOFF, $ALRATIOCUTOFF);

	# format results to output
	my $outputGene = $outputFile.'.gene.prediction';
	outputRawASARP($allAsarpsRef, 'ASARPgene', $outputGene);
	my $allNarOutput = formatOutputVerNAR($allAsarpsRef);</pre>
<a name='DESCRIPTION'></a><h1>DESCRIPTION</h1>
<p>The methodology in detail is explained in the paper:</p>
<p>Li G, Bahn JH, Lee JH, Peng G, Chen Z, Nelson SF, Xiao X. Identification of allele-specific alternative mRNA processing via transcriptome sequencing, Nucleic Acids Research, 2012, 40(13), e104</p>
<p>and its Supplementary Materials</p>
<p><img src='img/demo.jpeg' border=0></p>
<p>See <a href='http://nar.oxfordjournals.org/content/40/13/e104' target='_blank'>nar.oxfordjournals.org</a> for more details.</p>
<a name='SEE ALSO'></a><h1>SEE ALSO</h1>
<p><i><a href='fileParser.html' target='_blank'>fileParser</a></i>, <i><a href='snpParser.html' target='_blank'>snpParser</a></i>, <i><a href='MyConstants.html' target='_blank'>MyConstants</a></i></p>
<a name='COPYRIGHT'></a><h1>COPYRIGHT</h1>
<p>This pipeline is free software; you can redistribute it and/or modify it given that the related works and authors are cited and acknowledged.</p>
<p>This program is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose.</p>
<a name='AUTHOR'></a><h1>AUTHOR</h1>
<p>Cyrus Tak-Ming CHAN</p>
<p>Xiao Lab, Department of Integrative Biology &amp; Physiology, UCLA</p>
</div></body></html>
